src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -492,6 +492,355 @@ TEST(Aktualizr, FullWithUpdatesNeedReboot) {
+    auto received_event_it = received_events_.begin();
+
+    for (auto& expected_event : expected_event_order) {
+      auto received_event = *received_event_it;
+      if (expected_event != received_event) {
+        result = false;
+        break;
+      }
+
+      if (received_event == "DownloadProgressReport") {
+        while (*(++received_event_it) == "DownloadProgressReport")
+          ;
+      } else {
+        ++received_event_it;
+      }
+    }
+    return result;
+  }
+
+ private:
+  std::function<void(std::shared_ptr<event::BaseEvent>)> functor_;
+  std::vector<std::string> received_events_ = {};
+};
+
+/*
+ * Initialize -> UptaneCycle -> download updates and install them ->
+ * -> reboot emulated -> Initialize -> Fail installation finalization
+ *
+ * Verifies whether the uptane client is not at pending state after installation finalization failure
+ *
+ * Checks actions:
+ *
+ * - [x] Update is not in pending state anymore after failed finalization
+ */
+TEST(Aktualizr, FinalizationFailure) {
+  TemporaryDirectory temp_dir;
+  auto http_server_mock = std::make_shared<HttpInstallationFailed>(temp_dir.Path());
+  Config conf = UptaneTestCommon::makeTestConfig(temp_dir, http_server_mock->tls_server);
+  conf.pacman.fake_need_reboot = true;
+  conf.uptane.force_install_completion = true;
+  conf.uptane.polling_sec = 0;
+  conf.bootloader.reboot_sentinel_dir = temp_dir.Path();
+  auto storage = INvStorage::newStorage(conf.storage);
+
+  std::vector<std::string> expected_event_order = {
+      "SendDeviceDataComplete", "UpdateCheckComplete",    "DownloadProgressReport", "DownloadTargetComplete",
+      "DownloadProgressReport", "DownloadTargetComplete", "AllDownloadsComplete",   "InstallStarted",
+      "InstallTargetComplete",  "InstallStarted",         "InstallTargetComplete",  "AllInstallsComplete"};
+
+  std::vector<std::string> expected_report_order = {
+      "EcuDownloadStarted",     "EcuDownloadCompleted",   "EcuDownloadStarted",     "EcuDownloadCompleted",
+      "EcuInstallationStarted", "EcuInstallationApplied", "EcuInstallationStarted", "EcuInstallationCompleted"};
+
+  const std::string primary_ecu_id = "CA:FE:A6:D2:84:9D";
+  const std::string secondary_ecu_id = "secondary_ecu_serial";
+
+  {
+    // get update, install them and emulate reboot
+    Aktualizr aktualizr(conf, storage, http_server_mock);
+    EventHandler event_hdlr(aktualizr);
+    aktualizr.Initialize();
+
+    // verify currently installed version
+    std::vector<Uptane::Target> installed_versions;
+    size_t current_version{0};
+    size_t pending_version{0};
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+
+    // for some reason there is no any installed version at initial Aktualizr boot/run
+    // IMHO it should return currently installed version
+    ASSERT_TRUE(installed_versions.empty());
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
```

Comment:
Historically this has been because this includes Uptane metadata that we don't have available until we've communicated with the server. We could consider storing that metadata the first time that we receive Targets metadata. However, that is out of scope for this. Feel free to make a ticket for that, though.

Reply:
What about storing metadata on the first boot, i.e. store metadata about packages that are pre-installed. In case of OSTree, an image does have 'installed package' which  is ostree-managed rootfs prior to communication with server. I am just thinking loudly here, not sure if there is any practical need in storing "installed versions" an initial boot time though :)

Reply:
The problem is we don't have the Uptane metadata at first boot. We could consider trying to get it via garage-push/garage-sign somehow during bitbake, but that's a lot of work for something that doesn't seem critical. You are still correct that we should do _something_ to prevent installing something already installed, but pre-loading the metadata is probably unrealistic.